### PR TITLE
fallback to Gravatar for rendering avatar icons

### LINKED ISSUE #821: Avatars (or all images?) should go through proxy/authentication and be cached

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #821 

This changes up the strategy for rendering avatars in a few ways, because who knew rendering an image on screen could be so _fascinating_.

 - [x] initially render Hubot avatar while fetching an image
 - [x] if no GitHub avatar found, generate and try for a Gravatar avatar
 - [x] transform these into object URLs that the browser handles and can cache fine
 - [x] handle cache failures instead of retrying each time
 - [x] test against GHE setup and confirm the fallbacks make sense
   - [x] unauthenticated GHE requests will redirect and return something like `text/html` - handle this and skip it
   - [x] blacklist GHE avatars for now so we don't `HTTP 429` ourselves
   - [x] commit form user should change to use gravatar, instead of falling back to Hubot
   - [x] accounts list should render _something_ for the avatar
 - [x] test against non-GitHub repository
   - [x] don't show Hubot, do show Gravatars

### CODE REVIEW

File: app/src/ui/lib/avatar-in-memory-cache.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    // catch and ignore any network errors
+  }
+
+  inMemoryCache.set(requestUrl, url)
+
+  return url
+}
+
+/**
+ * Fetch an avatar associated to associate with a user, using graceful fallback
```
Comment (j-f1) [inline app/src/ui/lib/avatar-in-memory-cache.ts]: `associated to associate` 🤔❓

Comment (shiftkey) [reply app/src/ui/lib/avatar-in-memory-cache.ts]: 👍 What even are words?

**Comment by shiftkey** [comment]
Visually, this is what our current `master` does when you're only signed into GitHub Enterprise:

### GitHub Enterprise repository



### Some Other repository (in this case it's Desktop)



With this change, things get a bit more personal:

### GitHub Enterprise repository



### Some Other repository



This solution involves a fair bit of blacklisting of GitHub Enterprise avatars 😢  I have ideas for rolling this back later but it requires us to know about the version of GitHub Enterprise you're talking to - a job for a follow-up PR.

File: app/src/ui/history/commit-list.tsx [hunk 2]
```diff
@@ -32,13 +33,15 @@ export class CommitList extends React.Component<ICommitListProps, {}> {
 
     const gitHubUser =
       this.props.gitHubUsers.get(commit.author.email.toLowerCase()) || null
-    let avatarUser = null
-    if (gitHubUser) {
```
Comment (shiftkey) [inline app/src/ui/history/commit-list.tsx]: In a few places we were only rendering the avatar when we had an associated GitHub user. By grabbing the Gravatar URL when we can't find these users, the experience with non-GitHub repositories should be much nicer...

**Comment by iAmWillShepherd** [comment]
Looks good and works on my machine 🤞

**Pull request merged**